### PR TITLE
ENH: Slightly improve error when gufunc axes has wrong size

### PR DESCRIPTION
**Pull request by seberg**
My hope was to tweak it into something useful that:

    a @= b

can raise when `b` should have two dimensions and has two axes specified but actually only has one.
I didn't succeed, but I still think it a slight improvement to give the ufunc name and the actual core dimensions.

### CODE REVIEW

File: numpy/core/src/umath/ufunc_object.c [hunk 1]
```diff
@@ -1847,8 +1843,11 @@ _parse_axes_arg(PyUFuncObject *ufunc, int op_core_num_dims[], PyObject *axes,
             }
         }
         else {
-            PyErr_Format(PyExc_TypeError, "axes item %d should be a tuple",
-                         iop);
+            PyErr_Format(PyExc_TypeError,
+                    "%s: operand %d has %d core dimensions, "
+                    "but the axes item is not tuple.  (A single integer is "
+                    "also allowed when only one dimension is required.)",
```
Comment (mhvk) [inline numpy/core/src/umath/ufunc_object.c:1849]: I think I'd leave out the parenthetical sentence here, or write it as "A single integer is only allowed when just one dimension is required" -- but I think it does not necessarily help, and people should just read the documentation if confused.

**Review by mhvk — APPROVED** [review]
Yes, good to be more explicit!  One super nitpicky comment; can ignore if you wish.

File: numpy/core/src/umath/ufunc_object.c (hunk 1 above)
Comment (seberg) [reply numpy/core/src/umath/ufunc_object.c]: Yeah, agreed. Removed.  Something doesn't quite feel ideal still, but not sure what :), and I think its an improvement either way.

**Comment by seberg** [comment]
@mhvk one thing I noticed looking here, is that if core axes are optional, you cannot omit them by passing a shorter axes tuple.  I.e. `matul(2D, 2D, axes=[(-2, -1)], (-1,), (-1))` doesn't work.  Just curious if it is intentional, or because it would be complicated to allow?

**Comment by mhvk** [comment]
@seberg - I think I introduced the ability to give axes just before we allowed for optional axes, so I think it is more a question of oversight...

It might actually be quite neat if one could indeed use `axes` to signal one really wanted `matrix-stack @ vector-stack` rather than a stack of matrices. Though perhaps `(mat @ vec[..., np.newaxis]).squeeze(-1)` or just `np.einsum(...)` ends up being clearer. And the alternative of just implementing `matvec` and `vecmat` is perhaps cleanest of all.

**Comment by mhvk** [comment]
p.s. This seems all OK, so fine to merge.

**Comment by mattip** [comment]
Could these be `AxisError` instead of `ValueError` ?

**Comment by seberg** [comment]
Hmmm, right now we only use `AxisError` for an out-of-bounds axis I think, not for a different way `axis=` can be invalid.

So yeah, makes sense, although for myself I am happy to not broaden up `AxisError` for this (happy to do so I am just ±0 on it).

**Comment by mattip** [comment]
Something other than ValueError would be helpful in order to properly catch errors in ` a @= b`

**Comment by seberg** [comment]
Heh, I forgot about that idea...  I guess that is a good enough reason here, we might want to do it for more axis/axes related `ValueError`'s I guess.

**Comment by mattip** [comment]
This [is failing](https://github.com/numpy/numpy/actions/runs/3564647967/jobs/5988864709#step:4:3350). It used to raise `TypeError("axes item 2 should be a tuple")`.
```
a = np.arange(27.).reshape((3, 3, 3))
b = np.arange(10., 19.).reshape((3, 1, 3))
assert_raises(TypeError, inner1d, a, b, axes=[-1, -1, -1])
```

**Comment by seberg** [comment]
Grrr, restored the `TypeError` where it clearly is correct the long way around...  Not pretty, but maybe the practical path?

**Review by mhvk — APPROVED** [review]
Better logic indeed, even if a bit more tortuous.

**Review by mattip — APPROVED** [review]
LGTM. I added a release note, please take a look before merging.

**Pull request merged**